plugin/src/frontend.ml

```diff
@@ -34,27 +34,43 @@ let maybe_prove_coherence n inv_n idx_n : unit =
     Printf.printf "Defined coherence proof %s\n\n" (Id.to_string coh_n)
   else
     ()
-      
+
 (*
- * If the option is enabled, then prove section and retraction after find_ornament is called.
- * Otherwise, do nothing.
+ * If the option is enabled, then prove section, retraction, and adjunction after
+ * find_ornament is called. Otherwise, do nothing.
  *)
 let maybe_prove_equivalence n inv_n : unit =
+  let define_proof suffix ?(adjective=suffix) evd term =
+    let ident = with_suffix n suffix in
+    let const = define_term ident evd term true |> destConstRef in
+    Printf.printf "Defined %s proof %s\n\n" adjective (Id.to_string ident);
+    const
+  in
   if is_search_equiv () then
     let env = Global.env () in
     let evd = Evd.from_env env in
     let (promote, forget) = map_tuple make_constant (n, inv_n) in
     let l = initialize_lifting env evd promote forget in
     let (section, retraction) = prove_equivalence env evd l in
-    let sec_n = with_suffix n "section" in
-    let _ = define_term sec_n evd section true in
-    Printf.printf "Defined section proof %s\n\n" (Id.to_string sec_n);
-    let rec_n = with_suffix n "retraction" in
-    let _ = define_term rec_n evd retraction true in
-    Printf.printf "Defined retraction proof %s\n\n" (Id.to_string rec_n)
+    let sect = define_proof "section" evd section in
+    let retr0 = define_proof "retraction0" evd retraction ~adjective:"pre-retraction" in
```

Comment:
I would keep this one named as "retraction." Then rename the adjointify version to "retraction_adjointify" or something. Then, to get back a type you want for adjunction, pass the type explicitly to refer to the original retraction.  Though, if that is too hard for now, it is OK to just manually check that type in the tests.

Reply:
Are you sure about this naming scheme? The user should basically never touch the original retraction proof if they want to use adjunction.

I might not be following your point about the adjunction proof's type, but if I am, then that's what we already do. The inferred type for the adjunction proof is in terms of the original retraction proof (`${ornament}_retraction0`) wrapped with the adjointify lemma (`fg_id'`). Ideally, the user should just work with the adjointified retraction proof, for which the adjunction lemma has exactly the desired type.

Reply:
Yes, I am sure

Reply:
Done. The adjointified retraction proof is now called `${ornament}_retraction_adjoint`.